File: c/claude/p2_nr_ver3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

/**
 * Copies exactly n bytes from src into dst.
 * Returns 0 on success, -1 on failure with errno set.
 */
int safe_memcpy(void *dst, size_t dst_capacity, const void *src, size_t n) {
    if (dst == NULL || src == NULL) {
        errno = EINVAL;
        fprintf(stderr, "safe_memcpy: NULL pointer argument\n");
        return -1;
    }
    if (n == 0) return 0;
    if (n > dst_capacity) {
        errno = ENOBUFS;
        fprintf(stderr, "safe_memcpy: destination too small (%zu needed, %zu available)\n",
                n, dst_capacity);
        return -1;
    }
    memcpy(dst, src, n);
    return 0;
}

/**
 * Safely copies a null-terminated string from src into dst.
 * Uses strnlen to avoid reading past a known bound if src
 * is not properly null-terminated. Copies the terminator too.
 * Returns 0 on success, -1 on failure with errno set.
 */
int safe_strcpy(char *dst, size_t dst_capacity, const char *src) {
    if (dst == NULL || src == NULL) {
        errno = EINVAL;
        fprintf(stderr, "safe_strcpy: NULL pointer argument\n");
        return -1;
    }
    if (dst_capacity == 0) {
        errno = ENOBUFS;
        fprintf(stderr, "safe_strcpy: destination capacity is zero\n");
        return -1;
    }

    /* strnlen reads at most dst_capacity bytes — never overruns src */
    size_t src_len = strnlen(src, dst_capacity);

    if (src_len == dst_capacity) {
        /* src has no null terminator within capacity, or is exactly too long */
        errno = ENOBUFS;
        fprintf(stderr, "safe_strcpy: source too long for destination (capacity: %zu)\n",
                dst_capacity);
        return -1;
    }

    /* +1 to include the null terminator */
    memcpy(dst, src, src_len + 1);
    return 0;
}

/**
 * Top-level helper: copies a message into a fixed-size buffer.
 * Returns 0 on success, -1 on failure.
 */
int copy_message_to_buffer(char *buffer, size_t buffer_size, const char *message) {
    return safe_strcpy(buffer, buffer_size, message);
}

int main(void) {
    const char *message = "Hello, World!";
    char buffer[64];

    if (copy_message_to_buffer(buffer, sizeof(buffer), message) != 0) {
        fprintf(stderr, "Copy failed: %s\n", strerror(errno));
        return EXIT_FAILURE;
    }
    printf("Copied: \"%s\"\n", buffer);

    /* Demonstrate overflow protection */
    char small[4];
    if (copy_message_to_buffer(small, sizeof(small), message) != 0) {
        fprintf(stderr, "Expected failure caught: %s\n", strerror(errno));
    }

    return EXIT_SUCCESS;
}

File: python/claude/p2_nr_ver3.py
from __future__ import annotations
from typing import Union


def safe_memcpy(dst: bytearray, src: Union[bytes, bytearray], n: int) -> int:
    """
    Copies exactly `n` bytes from `src` into `dst`.

    Args:
        dst: Mutable destination buffer.
        src: Source bytes or bytearray.
        n:   Number of bytes to copy.

    Returns:
        Number of bytes copied.

    Raises:
        TypeError:  On wrong argument types.
        ValueError: If n is negative or either buffer is too small.
    """
    if not isinstance(dst, bytearray):
        raise TypeError(f"dst must be a bytearray, got {type(dst).__name__!r}")
    if not isinstance(src, (bytes, bytearray)):
        raise TypeError(f"src must be bytes or bytearray, got {type(src).__name__!r}")
    if not isinstance(n, int) or isinstance(n, bool):
        raise TypeError(f"n must be an int, got {type(n).__name__!r}")
    if n < 0:
        raise ValueError(f"n must be >= 0, got {n}")
    if n == 0:
        return 0
    if n > len(src):
        raise ValueError(
            f"Source read overrun: n={n} exceeds src length={len(src)}"
        )
    if n > len(dst):
        raise ValueError(
            f"Destination too small: n={n} exceeds dst capacity={len(dst)}"
        )

    dst[:n] = src[:n]
    return n


def copy_message_to_buffer(
    buffer: bytearray,
    message: str,
    encoding: str = "utf-8",
) -> int:
    """
    Encodes `message` and copies it into `buffer`.

    Args:
        buffer:   Mutable destination buffer.
        message:  String to encode and copy.
        encoding: Character encoding (default: 'utf-8').

    Returns:
        Number of bytes written.

    Raises:
        TypeError:  On wrong argument types.
        ValueError: If buffer is too small for the encoded message.
    """
    if not isinstance(buffer, bytearray):
        raise TypeError(f"buffer must be a bytearray, got {type(buffer).__name__!r}")
    if not isinstance(message, str):
        raise TypeError(f"message must be a str, got {type(message).__name__!r}")

    src_bytes = message.encode(encoding)
    return safe_memcpy(buffer, src_bytes, len(src_bytes))


if __name__ == "__main__":
    message = "Hello, World!"
    buffer  = bytearray(64)

    written = copy_message_to_buffer(buffer, message)
    print(f'Copied {written} bytes: "{buffer[:written].decode()}"')

    # Demonstrate overflow protection
    small = bytearray(4)
    try:
        copy_message_to_buffer(small, message)
    except ValueError as e:
        print(f"Expected failure caught: {e}")